File: secretsantacycles.cpp
#include <bits/stdc++.h>

using namespace std;

typedef vector<bool> vb;
typedef vector<int> vi;
typedef vector<vi> vvi;

void DFS(int s, vvi &components, vb &visited, vvi &adjList) {
    visited[s] = true;
    vi lst,frontier;
    lst.push_back(s); frontier.push_back(s);
    while (!frontier.empty()) {
        int u = frontier.back(); frontier.pop_back();
        for (int v: adjList[u]) {
            if (!visited[v]) {
                visited[v] = true;
                lst.push_back(v);
                frontier.push_back(v);
            }
        }
    }
    components.push_back(lst);
}

int main() {
    int N; scanf("%d",&N);
    vi adjList(N,0), indeg(N,0); vvi UDadjList(N,vi{});
    for (int i = 0; i < N; i++) {
        int a; scanf("%d",&a);
        adjList[i] = --a; indeg[a]++;
        UDadjList[i].push_back(a); UDadjList[a].push_back(i);
    }
    vvi components; vb visited(N,false);
    for (int i = 0; i < N; i++) {
        if (!visited[i]) DFS(i,components,visited,UDadjList);
    }
    int assignments = 0;
    for (vi lst: components) {
        int counter = 0;
        for (int u: lst) {
            if (indeg[u] > 1) counter += indeg[u] - 1;
        }
        assignments += components.size() > 1 ? max(counter,1) : counter;
    }
    printf("%d\n",assignments);
    return 0;
}

File: secretsantacycles.py
from sys import stdin,stdout

#treat graph as undirected, and count number of components, C
#number of assignments within each component = sum(indeg[i] - 1 for i in lst if indeg[i] > 1)
def DFS(s,visited,components,adjList):
    visited[s] = True
    lst,frontier = [s],[s]
    while frontier:
        u = frontier.pop()
        for v in adjList[u]:
            if not visited[v]:
                visited[v] = True
                lst.append(v)
                frontier.append(v)
    components.append(lst)

N = int(stdin.readline())
#adjList[i] contains the node that i is pointing towards
adjList,UDadjList,indeg = [-1 for _ in range(N)],[[] for _ in range(N)],[0 for _ in range(N)]
for i in range(N):
    a = int(stdin.readline()) - 1
    adjList[i] = a; indeg[a] += 1
    UDadjList[i].append(a); UDadjList[a].append(i)
#count number of components in O(N) time
components,visited = [],[False for _ in range(N)]
for i in range(N):
    if not visited[i]:
        DFS(i,visited,components,UDadjList)
assignments = 0
for lst in components:
    counter = sum(indeg[i] - 1 for i in lst if indeg[i] > 1)
    # add to assignments by minimum of 1 if there are at least two components
    assignments += max(counter,1) if len(components) > 1 else counter
stdout.write(f"{assignments}\n") 
